File: pitstop_email_digest/pitstop_email_digest/report/aged_pending_purchase_orders/aged_pending_purchase_orders.py
# Copyright (c) 2026, QCS and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate, nowdate


def execute(filters=None):
    filters = frappe._dict(filters or {})
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {
            "label": _("Purchase Order"),
            "fieldname": "purchase_order",
            "fieldtype": "Link",
            "options": "Purchase Order",
            "width": 180,
        },
        {
            "label": _("Supplier"),
            "fieldname": "supplier",
            "fieldtype": "Link",
            "options": "Supplier",
            "width": 160,
        },
        {
            "label": _("Supplier Name"),
            "fieldname": "supplier_name",
            "fieldtype": "Data",
            "width": 220,
        },
        {
            "label": _("Grand Total"),
            "fieldname": "grand_total",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 140,
        },
        {
            "label": _("Age (Days)"),
            "fieldname": "age",
            "fieldtype": "Int",
            "width": 110,
        },
    ]


def get_data(filters):
    conditions = []
    values = {
        "as_on_date": getdate(filters.get("as_on_date") or nowdate()),
        "min_age": int(filters.get("min_age") or 90),
    }

    if filters.get("company"):
        conditions.append("po.company = %(company)s")
        values["company"] = filters.get("company")

    if filters.get("supplier"):
        conditions.append("po.supplier = %(supplier)s")
        values["supplier"] = filters.get("supplier")

    conditions_str = " and ".join(conditions)
    if conditions_str:
        conditions_str = "and " + conditions_str

    return frappe.db.sql(
        f"""
            select
                po.name as purchase_order,
                po.supplier,
                po.supplier_name,
                po.grand_total,
                po.currency,
                datediff(%(as_on_date)s, po.transaction_date) as age
            from `tabPurchase Order` po
            where
                po.docstatus = 1
                and po.status not in ('Closed', 'Completed')
                and datediff(%(as_on_date)s, po.transaction_date) > %(min_age)s
                {conditions_str}
            order by age desc, po.transaction_date asc
        """,
        values,
        as_dict=1,
    )

File: pitstop_email_digest/pitstop_email_digest/report/aged_pending_purchase_orders/aged_pending_purchase_orders.js
// Copyright (c) 2026, QCS and contributors
// For license information, please see license.txt

frappe.query_reports["Aged Pending Purchase Orders"] = {
	filters: [
		{
			fieldname: "as_on_date",
			label: __("As On Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "min_age",
			label: __("Minimum Age (Days)"),
			fieldtype: "Int",
			default: 90,
			reqd: 1,
		},
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "supplier",
			label: __("Supplier"),
			fieldtype: "Link",
			options: "Supplier",
		},
	],

	onload: function (report) {
		report.page.add_inner_button(__("Close"), function () {
			let data = frappe.query_report.get_checked_items() || [];
			let names = [...new Set(data.map((d) => d.purchase_order).filter(Boolean))];

			if (!names.length) {
				frappe.msgprint(__("Please select at least one Purchase Order."));
				return;
			}

			frappe.confirm(
				__("Are you sure you want to close {0} Purchase Order(s)?", [names.length]),
				function () {
					frappe.call({
						method: "erpnext.buying.doctype.purchase_order.purchase_order.close_or_unclose_purchase_orders",
						args: {
							names: JSON.stringify(names),
							status: "Closed",
						},
						freeze: true,
						freeze_message: __("Closing Purchase Orders..."),
						callback: function () {
							frappe.show_alert({
								message: __("{0} Purchase Order(s) closed", [names.length]),
								indicator: "green",
							});
							frappe.query_report.refresh();
						},
					});
				}
			);
		});
		frappe.breadcrumbs.add("Buying");
	},

	get_datatable_options(options) {
		return Object.assign(options, {
			checkboxColumn: true,
		});
	},
};
